src/transaction/log_append.cpp

```diff
@@ -1569,6 +1571,157 @@ prior_is_tde_encrypted (const log_prior_node *node)
   return node->tde_encrypted;
 }
 
+static size_t
+prior_node_serialized_size (const log_prior_node &node)
+{
+  size_t total_size = 0;
+
+  // fixed data
+  total_size += sizeof (node.log_header)
+		+ sizeof (node.start_lsa)
+		+ sizeof (node.tde_encrypted)
+		+ sizeof (node.data_header_length)
+		+ sizeof (node.ulength)
+		+ sizeof (node.rlength);
+
+  // variable data
+  total_size += static_cast<size_t> (node.data_header_length);
+  total_size += static_cast<size_t> (node.ulength);
+  total_size += static_cast<size_t> (node.rlength);
+
+  return total_size;
+}
+
+static size_t
+prior_list_serialized_size (const log_prior_node *head)
+{
+  size_t total_size = 0;
+  for (const log_prior_node *nodep = head; nodep != nullptr; nodep = nodep->next)
+    {
+      total_size += prior_node_serialized_size (*nodep);
+    }
+  return total_size;
+}
+
+static void
+prior_node_serialize (const log_prior_node &node, std::string &serialized)
+{
+#define copy_mem(mem, memsz) if ((memsz) > 0) serialized.append (mem, memsz)
+#define copy(val) copy_mem (reinterpret_cast<const char *> (&(val)), sizeof (val))
+
+  copy (node.log_header);
+  copy (node.start_lsa);
+  copy (node.tde_encrypted);
+  copy (node.data_header_length);
+  copy (node.ulength);
+  copy (node.rlength);
+
+  copy_mem (node.data_header, node.data_header_length);
+  copy_mem (node.udata, node.ulength);
+  copy_mem (node.rdata, node.rlength);
+
+#undef copy_mem
+#undef copy
+}
+
+static const char *
+prior_node_deserialize (const char *ptr, log_prior_node &node)
+{
+#define copy_mem(mem, memsz) std::memcpy (mem, ptr, memsz); ptr += (memsz)
```

Comment:
Nitpicking, couldn't these macros be some [templated, if needed] inlined functions?

Reply:
I considered it... But macros were preferred since there are very specific needs here (append mem/object to string) and they allowed the lazy me to use less arguments :)

I also considered using lambdas, but I am not sure the compiler will know to inline.

Do you think the inlined/template functions will improve readability?

Reply:
Lambdas with this complexity would be inlined - so they say - but generic (ie: auto argument) lambdas are`c++14`.
Templates would add explicitness but also a lot of verbosity.
Sticking to the [not] good ol' macros...